src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/UntypedReferenceRewriter.cs

```diff
@@ -0,0 +1,108 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// For expressions over a reference search parameter where the type is not specified
+    /// (e.g. "abc" instead of "Patient/123") AND the search parameter is constrained to be
+    /// of only one type, rewrites the expression to specify the single target type.
+    /// </summary>
+    internal class UntypedReferenceRewriter : ExpressionRewriterWithInitialContext<object>
+    {
+        public static readonly UntypedReferenceRewriter Instance = new UntypedReferenceRewriter();
+
+        public override Expression VisitSearchParameter(SearchParameterExpression expression, object context)
+        {
+            // Handle reference search parameters as well as composite search parameters with one or more reference components.
+            // We fist create a bitmask with bits set representing the component indexes that are candidates for this rule.
```

Comment:
>bitmask with bits set representing the component indexes that are candidates for this rule [](start = 32, length = 90)

Any chance I can persuade choice of BitArray over int with bit operations?
I understand BitArray would bring performance downgrade, but there is chance it would be easier to read...

Reply:
I don't know. Would `componentCandidates.And(componentsPresent.Not())` really improve readability :)?

Reply:
Not that case, but rest of them, I personally always forget which shift left or shift right is increasing value, and which one is decreasing (without context of overflow), access to bit by index is much easier in case of BitArray.
But someone can argue should I even be allowed to read code if I don't have such a basic bit operation understanding :D 
Again, it's just an opinion.

---
In reply to: [538869133](https://github.com/microsoft/fhir-server/pull/1497#discussion_r538869133) [](ancestors = 538869133)

Reply:
Added more comments.